strings.ToLower() on owner

cjmartian opened the pull request:
<!-- Thank you for contributing a pull request! Here are a few tips to help you:

1. If this is your first contribution, make sure you've read the Contribution Checklist https://developers.mattermost.com/contribute/getting-started/contribution-checklist/
2. Read our blog post about "Submitting Great PRs" https://developers.mattermost.com/blog/2019-01-24-submitting-great-prs
3. Take a look at other repository specific documentation at https://developers.mattermost.com/contribute
-->

#### Summary
<!--
A description of what this pull request does.
-->
Lowers orgs/owners when trying to fetch repos.
#### Ticket Link
<!--
If this pull request addresses a Help Wanted ticket, please link the relevant GitHub issue, e.g.

  Fixes https://github.com/mattermost/mattermost-server/issues/XXXXX

Otherwise, link the JIRA ticket.
-->
https://github.com/mattermost/mattermost-plugin-github/issues/377

mattermod commented:
Hello @cjmartian,

Thanks for your pull request! A Core Committer will review your pull request soon. For code contributions, you can learn more about the review process [here](https://developers.mattermost.com/contribute/getting-started/code-review/).

cjmartian commented:
This may not be all that is needed. I think it makes sense to add a test for this.

jasonblais commented:
@cjmartian A test would be great to add :+1:

Once you've added one or if you have any questions, let me know :)

cjmartian commented:
@jasonblais You got it! I'll poke around.

cjmartian commented:
@jasonblais alright, I added what I think is a better test. I've also changed some of the functionality of `command.go` which I'm not sure was appropriate. In either case, the test I provided might better test the subscribe functionality at the cost of using `strings.Contains()` instead of `strings.HasPrefix()` for confirming that something is a label.

Let me know if this seems alright. If not, I can go back and take a different approach altogther.

On another note, the tests will fail because of my styling on this line:

```
	if valid && ( hasLabel || isRepo ) {
```

If I could get some guidance on styling, that would be great. This is my first real exposure to Golang.

Thanks!

jasonblais commented:
Thank you @cjmartian! 

I've requested reviews from @hanzei and @larkox who can better guide you on above (both the styling and the test case).

> This is my first real exposure to Golang.

I'm so glad to hear that! Hope it's been a good learning experience so far @cjmartian :)

larkox reviewed (approved):
LGTM, just a quesiton.

cjmartian commented on server/plugin/command.go:
```diff
@@ -44,21 +46,26 @@ func validateFeatures(features []string) (bool, []string) {
+	isRepo := false
 	for _, f := range features {
 		if _, ok := validFeatures[f]; ok {
 			continue
 		}
-		if strings.HasPrefix(f, "label") {
+		if strings.Contains(f, "label") {
```
Specifically for this test case: https://github.com/mattermost/mattermost-plugin-github/pull/380/files#diff-05e1a5ecb4a0133c535cadfdf50ce3daR75-R84

I'm open to suggestions, but I think we need to use Contains() here because otherwise when a test passes something like `{"subscribe", "dhaussermann/hello-world", issues,label:"help"}` it won't pick up on the label.

larkox replied:
In theory, the features will be separated by commas, so in the example you are giving, f will be `issues` or `label:"help"`. In case of `label:"help"` the HasPrefix should still work. Am I right?

cjmartian replied:
I can try to split the arguments into different args, but from what I saw in the test the argument was `issues,label:help` as one argument. I can split that argument into two different args effectively making it `issues` then `label:help` which will allow us to use `HasPrefix()`. 

I just didn't want to alter the test too much just in case it was catching some particular functionality. 

Let me know! If you think that's better, I can absolutely change it! 🙂

cjmartian replied:
I changed the test locally, if we think this is the approach we want to take let me know and I can push it up!

larkox replied:
After a second look, many of the things you are checking here probably should not go here. The repo name should never reach the validateFeatures function. It will be handled in "handleSubscribesAdd", and only the "option list" will be passed to the validate features function. That is the main reason why the tests needed some tweaking too.

cjmartian replied:
Ah! I'll take another look then. Thank you for explaining that to me!

hanzei commented on server/plugin/subscriptions.go:
```diff
@@ -101,6 +101,8 @@ func (p *Plugin) Subscribe(ctx context.Context, githubClient *github.Client, use
 		return errors.Errorf("invalid repository")
 	}
 
+	owner = strings.ToLower(owner)
```
Did you test these changes with a repo that has a lower case letter it it?

jfrerich replied:
Hi @cjmartin.  It appears your PR is really close.  Can you address @hanzei's question?

hanzei reviewed (changes requested):
Thanks for taking a look. I just want to confirm you tested the changes

mattermod commented:
This PR has been automatically labelled "stale" because it hasn't had recent activity.
A core team member will check in on the status of the PR to help with questions.
Thank you for your contribution!

/cc @jasonblais @jfrerich

larkox commented:
@cjmartian Are you able to continue working on this PR?

cjmartian commented:
Yeah, I can. Sorry I've been exceptionally busy lately. 

I can take a look later this week/weekend

hanzei commented:
@cjmartian Heads up that I took over the PR given that it was inactive for some time now.

cjmartian commented:
No worries @hanzei 
Sorry about letting this go stale, I appreciate you finishing it off.

mattermod commented:
This PR has been automatically labelled "stale" because it hasn't had recent activity.
A core team member will check in on the status of the PR to help with questions.
Thank you for your contribution!

/cc @jasonblais @jfrerich @emilyacook

jasonblais commented:
@DHaussermann looks like this one is ready for testing

DHaussermann commented:
Hi @cjmartian Sorry your PR does not seem to be testable right now. I am seeing some regressions in the modals when trying to fetch repos in the `master` branch which would make it inconclusive to test this now. 

I have created https://github.com/mattermost/mattermost-plugin-github/issues/451 to address this.
cc @hanzei

hanzei commented:
/update-branch

DHaussermann commented:
Thanks for the PR @cjmartian I have done some testing and I'm not sure this is a complete solution. When I add a new subscription now all letters are lower case however, these subscriptions are not functional when the casing is mismatched compared to the repository. 

Steps:
1. Locate a mixed case repo such as https://github.com/DHaussermann/hello-world
2. Add a subscription in s a channel using either `/github subscriptions add  DHaussermann/hello-world` OR  `/github subscriptions add  dhaussermann/hello-world`   
3. Note that subscription is created for `dhaussermann/hello-world` even if the `add` command had uppercase letters 
4. Test the subscription delivery by creating a new issue
**Observed**: No subscription event is delivered 

Can you see if you can reproduce this? It seems like the subscriptions won't work without the uppercase letters.

DHaussermann commented:
@cjmartian I believe the issue found has to be investigated before the next round of testing.
I'm adding the label for awaiting submitter action. Please reach out if I've misunderstood something.

mattermod commented:
This PR has been automatically labelled "stale" because it hasn't had recent activity.
A core team member will check in on the status of the PR to help with questions.
Thank you for your contribution!

/cc @jasonblais @jfrerich @emilyacook

DHaussermann commented:
/update-branch

DHaussermann commented:
Discussed with @dipak-demansol today. Once successful subscription deliveries from mixed case repos are confirmed working this would be good to merge.

dipak-demansol commented:
> Discussed with @dipak-demansol today. Once successful subscription deliveries from mixed case repos are confirmed working this would be good to merge.

after subscribe the repo of xyz channel, if i report or close any issue in the repo, then i notified in the channel but if i edit the same issue then i'm not notified so in that case it should notify me in the channel? or i can pass it?

dipak-demansol commented:
> > Discussed with @dipak-demansol today. Once successful subscription deliveries from mixed case repos are confirmed working this would be good to merge.
> 
> after subscribe the repo of xyz channel, if i report or close any issue in the repo, then i notified in the channel but if i edit the same issue then i'm not notified so in that case it should notify me in the channel? or i can pass it?

@DHaussermann LGTM

dipak-demansol commented:
> > > Discussed with @dipak-demansol today. Once successful subscription deliveries from mixed case repos are confirmed working this would be good to merge.
> > 
> > 
> > after subscribe the repo of xyz channel, if i report or close any issue in the repo, then i notified in the channel but if i edit the same issue then i'm not notified so in that case it should notify me in the channel? or i can pass it?
> 
> @DHaussermann LGTM

@DHaussermann as you told me there are more feature to test as per this PR so i will update you tomorrow.

dipak-demansol commented:
> @dipak-demansol as we discussed all subscriptions events do **not** need to be tested for this PR.
> If the subscription deliveries work for any event in a repo where the name is mixed case I would be satisfied. I do **not** see a risk here where some event types would no longer work.

@DHaussermann i tested again and found that with the mix case repo is subscribed but its not delivering the notifications.

dipak-demansol commented:
@hanzei @DHaussermann LGTM, no further testing required.
